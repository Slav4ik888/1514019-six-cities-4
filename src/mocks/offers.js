export default [
  {
    id: 1,
    isPremium: true,
    isFavourite: false,
    previewImage: `img/apartment-03.jpg`,
    price: 180,
    rating: 5,
    cardTitle: `Beautiful & luxurious apartment at great location`,
    offerType: `House`,
  },
  {
    id: 2,
    isPremium: true,
    isFavourite: false,
    previewImage: `img/apartment-01.jpg`,
    price: 180,
    rating: 3,
    cardTitle: `Nice, cozy, warm big bed apartment`,
    offerType: `Hotel`
  },
  {
    id: 3,
    isPremium: false,
    isFavourite: false,
    previewImage: `img/apartment-02.jpg`,
    price: 10,
    rating: 5,
    cardTitle: `Wood and stone place`,
    offerType: `Apartment`,
  },
  {
    id: 4,
    isPremium: false,
    isFavourite: false,
    previewImage: `img/apartment-01.jpg`,
    price: 180,
    rating: 5,
    cardTitle: `Beautiful & luxurious apartment at great location`,
    offerType: `Private room`,
  },
  {
    id: 5,
    isPremium: false,
    isFavourite: true,
    previewImage: `img/apartment-03.jpg`,
    price: 480,
    rating: 4,
    cardTitle: `Beautiful & luxurious apartment at great location`,
    offerType: `Hotel`
  },
  {
    id: 6,
    isPremium: false,
    isFavourite: false,
    previewImage: `img/apartment-01.jpg`,
    price: 680,
    rating: 3,
    cardTitle: `Nice, cozy, warm big bed apartment`,
    offerType: `House`,
  },
  {
    id: 7,
    isPremium: false,
    isFavourite: true,
    previewImage: `img/apartment-02.jpg`,
    price: 180,
    rating: 4,
    cardTitle: `Wood and stone place`,
    offerType: `Apartment`,
  },
  {
    id: 8,
    isPremium: true,
    isFavourite: false,
    previewImage: `img/apartment-01.jpg`,
    price: 180,
    rating: 4,
    cardTitle: `Beautiful & luxurious apartment at great location`,
    offerType: `Private room`,
  },
];
