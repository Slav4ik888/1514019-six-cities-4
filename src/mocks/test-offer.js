export const testOffer = {
  id: 1,
  isPremium: true,
  isFavourite: false,
  previewImage: `img/apartment-03.jpg`,
  pictures: [
    `img/apartment-01.jpg`,
    `img/apartment-02.jpg`,
    `img/apartment-03.jpg`,
    `img/room.jpg`,
    `img/studio-01.jpg`,
    `img/studio-photos.jpg`],
  amenities: [`Wifi`, `Heating`, `Kitchen`, `Cable TV`],
  bedrooms: 4,
  maxGuestsNumber: 5,
  description: `This cozy and complete apartment in the heart of New York is a typical, traditional and authentic feel for how New Yorkers live. I welcome all guests to stay at this place during their stay and visit of New York. 
  Please note that this apartment is on the sixth floor and there is no lift (elevator).
  The apartment is convenient for all guests, fitted with a kitchen, small dining area and two bedrooms – couples and groups of friends will find this accommodation most optimal.`,
  host: {
    photo: `img/avatar-angelina.jpg`,
    name: `Angelina Jonson`,
    super: true
  },
  price: 180,
  rating: 5,
  cardTitle: `Beautiful & luxurious apartment at great location`,
  offerType: `House`,
  coordinates: [48.856663, 2.351556],
  city: {
    zoom: 8,
    coordinates: [48.856663, 2.351556],
  },
  location: {
    zoom: 8,
  },
  reviews: [
    {
      id: 1,
      author: {
        photo: `img/avatar-angelina.jpg`,
        name: `Strange jober`
      },
      description: `small dining area and two bedrooms – couples and groups of friends will find this accommodation`,
      date: 1593095836000,
      rating: 3
    },
    {
      id: 2,
      author: {
        photo: ``,
        name: `Slava`
      },
      description: `dining area and five bedrooms – couples and groups of friends will find this accommodation`,
      date: 1593095936000,
      rating: 4.7
    },
  ],

};
