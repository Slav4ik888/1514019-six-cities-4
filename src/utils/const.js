export const offerTypes = {
  apartment: `Apartment`,
  room: `Private room`,
  house: `House`,
  hotel: `Hotel`,
};

export const cities = [
  `Paris`,
  `Cologne`,
  `Brussels`,
  `Amsterdam`,
  `Hamburg`,
  `Dusseldorf`,
];

export const citiesIdx = {
  Paris: 0,
  Cologne: 1,
  Brussels: 2,
  Amsterdam: 3,
  Hamburg: 4,
  Dusseldorf: 5,
};

export const coordsCities = [
  [48.856663, 2.351556],
  [50.930779, 6.938399],
  [50.851309, 4.351718],
  [52.38333, 4.9],
  [53.552645, 9.966287],
  [51.230569, 6.787428],
];


export const AppRoute = {
  SIGN_IN: `/login`,
  MAIN: `/`,
  // MAIN_EMPTY: `/main_empty`,
  ROOM: `/offer`,
  ROOM_ID: `/offer/:id`,
  FAVORITES: `/favorites`,
  // FAVORITES_EMPTY: `/favorites-empty`,
};

export const placesType = {
  NEAR: `NEAR`,
  CITY: `CITY`,
  FAVORITE: `FAVORITE`,
  OFFER_DETAILS: `OFFER_DETAILS`,
};

export const MAX_REVIEW_COUNT = 10;

export const sortType = {
  POPULAR: `popular`,
  LOW_HIGH: `low-high`,
  HIGH_LOW: `high-low`,
  RATING: `rating`,
};

export const pageType = {
  MAIN: `MAIN`,
  SIGN_IN: `SIGN_IN`,
  OFFER: `OFFER`,
  FAVORITES: `FAVORITES`,
  FAVORITES_EMPTY: `FAVORITES_EMPTY`,
};
