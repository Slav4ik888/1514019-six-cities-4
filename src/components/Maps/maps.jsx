import React, {createRef, PureComponent} from 'react';
import pt from 'prop-types';
import cl from 'classnames';
import {pageType} from '../../utils/const.js';
import {offerPropTypes} from '../../utils/prop-types-templates.js';

import L from 'leaflet';


class Maps extends PureComponent {
  constructor(props) {
    super(props);
    this._mapRef = createRef();
    this._renderMarker = this._renderMarker.bind(this);
    this._clearMap = this._clearMap.bind(this);
    this._map = null;

    this._zoom = 12;
    this._markers = []; // Массив для хранения маркеров выведенных на карту
  }

  componentDidMount() {
    const {activeCoords} = this.props;

    if (!this._mapRef.current) {
      return;
    }

    const city = activeCoords;

    // Инициализация карты
    this._map = L.map(this._mapRef.current, {
      center: city,
      zoom: this._zoom,
      zoomControl: false,
      marker: true
    });


    // Устанавливаем центр карты по координатам city
    this._map.setView(city, this._zoom);

    L
      .tileLayer(`https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
        attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
      })
      .addTo(this._map);

    this._renderMarker();
  }


  componentWillUnmount() {
    this.map = null;
  }


  componentDidUpdate() {
    this._clearMap();
    const {activeCoords} = this.props;

    const city = activeCoords;

    this._map.setView(city, this._zoom);

    this._renderMarker();
  }


  _clearMap() {
    // Чистим карту если есть старые маркеры
    if (this._markers) {
      this._markers.forEach((item) => {
        this._map.removeLayer(item);
      });

      this._markers = [];
    }
  }

  _renderMarker() {
    const {offers, activeOffer} = this.props;

    const icon = L.icon({
      iconUrl: `/img/pin.svg`,
      iconSize: [27, 33]
    });
    // Выводим маркеры offers на карту
    if (offers) {
      offers.forEach((offer) => {
        let marker = L.marker(offer.coordinates, {icon});
        this._markers.push(marker); // Сохраняем маркер, чтобы потом удалить его
        this._map.addLayer(marker);

        // return (L
        //   .marker(offer.coordinates, {icon})
        //   .addTo(this._map)
        // );
      });
    }

    // Маркер активного activeOffer
    if (activeOffer) {
      const activeIcon = L.icon({
        iconUrl: `/img/pin-active.svg`,
        iconSize: [27, 33],
      });
      let marker = L.marker(activeOffer.coordinates, {icon: activeIcon});
      this._markers.push(marker);
      this._map.addLayer(marker);
    }
  }


  render() {
    const {type} = this.props;

    const mapClasses = cl(`map`,
        {[`property__map`]: type === pageType.OFFER},
        {[`cities__map`]: type === pageType.MAIN}
    );

    return (
      <section className={mapClasses}>
        <div id="map" ref={this._mapRef} style={{height: `100%`}}></div>
      </section>
    );
  }
}

Maps.propTypes = {
  type: pt.oneOf([pageType.MAIN, pageType.OFFER]).isRequired,
  offers: pt.arrayOf(
      pt.shape(offerPropTypes).isRequired
  ).isRequired,
  activeCoords: pt.array.isRequired,
  activeOffer: pt.shape(offerPropTypes),
};


export default Maps;
