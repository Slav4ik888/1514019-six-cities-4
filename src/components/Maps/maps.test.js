import React from 'react';
import renderer from 'react-test-renderer';
import Maps from './maps.jsx';
import {pageType} from '../../utils/const.js';
import {coordsCities} from '../../utils/const.js';
import {offers} from '../../mocks/offers.js';
import {testOffer} from '../../mocks/test-offer.js';


it(`Render Snapshot <Maps/> pageType.MAIN`, () => {
  const component = renderer.create(
      <Maps
        type={pageType.MAIN}
        offers={offers.Paris}
        activeOffer={testOffer}
        activeCoords={coordsCities[0]}
      />
      , {
        createNodeMock: () => {
          return document.createElement(`div`);
        }
      }
  ).toJSON();

  expect(component).toMatchSnapshot();
});

it(`Render Snapshot <Maps/> pageType.OFFER`, () => {
  const component = renderer.create(
      <Maps
        type={pageType.OFFER}
        offers={offers.Paris}
        activeOffer={testOffer}
        activeCoords={coordsCities[0]}
      />
      , {
        createNodeMock: () => {
          return document.createElement(`div`);
        }
      }
  ).toJSON();

  expect(component).toMatchSnapshot();
});

// npm run test.jest -- -u maps.test.js
// npm test maps.test.js
