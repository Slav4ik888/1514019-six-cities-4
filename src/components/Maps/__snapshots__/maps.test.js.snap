// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Snapshot <Maps/> pageType.MAIN 1`] = `
<section
  className="map cities__map"
>
  <div
    id="map"
    style={
      Object {
        "height": "100%",
      }
    }
  />
</section>
`;

exports[`Render Snapshot <Maps/> pageType.OFFER 1`] = `
<section
  className="map property__map"
>
  <div
    id="map"
    style={
      Object {
        "height": "100%",
      }
    }
  />
</section>
`;
